File: android/sync/src/main/java/com/memorypets/sync/SyncOrchestrator.kt
package com.memorypets.sync

import com.memorypets.core.crypto.VaultCrypto
import com.memorypets.core.model.Envelope
import com.memorypets.core.model.SyncOutcome
import com.memorypets.sync.local.VaultBlobStore
import com.memorypets.sync.prefs.AppPrefs
import com.memorypets.sync.remote.CloudSyncApi
import com.memorypets.sync.remote.CloudAuthExpiredException
import com.memorypets.sync.remote.dto.PutVaultReq
import com.memorypets.sync.remote.dto.toDTO
import com.memorypets.sync.remote.dto.toDomain
import dagger.Lazy
import kotlinx.coroutines.flow.first
import retrofit2.HttpException
import javax.inject.Inject
import javax.inject.Singleton
import javax.crypto.AEADBadTagException

/**
 * 云同步编排器 —— 与 packages/host/lib/cloud-sync.mjs 的 push / pull / confirmVersion 语义对齐。
 *
 * 乐观并发流程：
 *   1) GET /vault -> remoteVersion
 *   2) remoteVersion > 本地 currentVersion -> 解密 remote envelope -> 采用（失败则 ConflictNeedManual）
 *   3) 本地有改动 (dirty flag) 或本地版本 0 -> PUT /vault (expectedVersion = lastReadVersion)
 *        200 -> confirmVersion(v+1)
 *        409 -> 读取 current.envelope -> 解密采用 -> 再 PUT 一次 expectedVersion=current.version
 *                 仍 409 -> ConflictNeedManual 交给用户
 */
@Singleton
class SyncOrchestrator @Inject constructor(
    private val api: Lazy<CloudSyncApi>,
    private val blobStore: VaultBlobStore,
    private val prefs: AppPrefs,
) {

    /**
     * @param masterPassword 当前解锁会话中的主密码（仅解密冲突时使用；调用方用完立即置 null）
     * @param buildLocalEnvelope 若本地有脏数据，调用此 lambda 得到最新密封好的 Envelope
     */
    suspend fun syncNow(
        masterPassword: CharArray?,
        buildLocalEnvelope: suspend () -> Envelope?
    ): SyncOutcome {
        val token = prefs.sessionToken.first()
            ?: return SyncOutcome.AuthExpired("not logged in")
        val bearer = "Bearer $token"
        val deviceId = prefs.deviceId.first()

        return try {
            // 1) GET remote
            val getResp = api.get().getVault(bearer)
            if (!getResp.isSuccessful || getResp.body()?.ok != true) {
                val code = getResp.code()
                if (code == 401) return SyncOutcome.AuthExpired("token rejected")
                return SyncOutcome.NetworkError(
                    RuntimeException("GET /vault failed HTTP $code")
                )
            }
            val body = getResp.body()!!
            val remoteVersion = body.version
            val remoteEnv = body.envelope

            var adoptedRemote = false

            // 2) 有远程密文 & 更新 -> 拉取 & 解密采用
            if (remoteEnv != null && remoteVersion > blobStore.currentVersion().first()) {
                if (masterPassword == null) {
                    return SyncOutcome.ConflictNeedManual(
                        theirVersion = remoteVersion,
                        myVersion = blobStore.currentVersion().first()
                    )
                }
                try {
                    VaultCrypto.unlock(remoteEnv.toDomain(), masterPassword) // 先验证能解开再落盘
                    blobStore.overwriteWithRemote(
                        envelope = remoteEnv.toDomain(),
                        newVersion = remoteVersion,
                        updatedAtMs = body.updatedAt ?: System.currentTimeMillis()
                    )
                    prefs.markSynced(remoteVersion, System.currentTimeMillis())
                    adoptedRemote = true
                } catch (e: AEADBadTagException) {
                    return SyncOutcome.DecryptFailed(
                        "remote vault sealed with different master password"
                    )
                }
            }

            // 3) 本地有脏改动 / 从未提交过 -> push
            val localDirty = blobStore.isDirty()
            val nothingOnRemote = remoteVersion == 0L && remoteEnv == null
            if (!localDirty && !nothingOnRemote) {
                return if (adoptedRemote)
                    SyncOutcome.Pulled(remoteVersion, remoteAdopted = true)
                else
                    SyncOutcome.Pushed(remoteVersion)
            }

            val myEnv = buildLocalEnvelope()
                ?: run {
                    return if (adoptedRemote) SyncOutcome.Pulled(remoteVersion, true)
                           else SyncOutcome.Pushed(remoteVersion)
                }

            val expectedVersion = maxOf(remoteVersion, blobStore.currentVersion().first())
            val firstPut = putVaultWithRetry(bearer, myEnv, expectedVersion, deviceId)
            if (firstPut is SyncOutcome.Pushed) {
                blobStore.confirmVersion(firstPut.newVersion, System.currentTimeMillis())
                prefs.markSynced(firstPut.newVersion, System.currentTimeMillis())
                return firstPut
            }
            // 409 -> retry once
            if (firstPut is SyncOutcome.ConflictNeedManual) {
                val theirVersion = firstPut.theirVersion
                val theirEnvDto = fetchCurrentEnvAfter409(bearer)?.envelope
                    ?: return SyncOutcome.NetworkError(RuntimeException("failed to refetch winner after 409"))
                if (masterPassword != null) {
                    try {
                        VaultCrypto.unlock(theirEnvDto.toDomain(), masterPassword)
                        blobStore.overwriteWithRemote(
                            theirEnvDto.toDomain(), theirVersion, System.currentTimeMillis()
                        )
                    } catch (_: AEADBadTagException) {
                        return SyncOutcome.DecryptFailed("different master password on 409 winner")
                    }
                }
                // retry PUT with the winner version
                val retry = putVaultWithRetry(bearer, myEnv, theirVersion, deviceId)
                return when (retry) {
                    is SyncOutcome.Pushed -> {
                        blobStore.confirmVersion(retry.newVersion, System.currentTimeMillis())
                        prefs.markSynced(retry.newVersion, System.currentTimeMillis())
                        SyncOutcome.Pushed(retry.newVersion)
                    }
                    else -> retry
                }
            }
            firstPut
        } catch (e: CloudAuthExpiredException) {
            SyncOutcome.AuthExpired(e.message ?: "session expired")
        } catch (e: HttpException) {
            if (e.code() == 401) SyncOutcome.AuthExpired("401")
            else SyncOutcome.NetworkError(e)
        } catch (t: Throwable) {
            SyncOutcome.NetworkError(t)
        }
    }

    private suspend fun putVaultWithRetry(
        bearer: String,
        env: Envelope,
        expectedVersion: Long,
        deviceId: String
    ): SyncOutcome {
        val req = PutVaultReq(
            envelope = env.toDTO(),
            expectedVersion = expectedVersion,
            deviceId = deviceId
        )
        val resp = runCatching { api.get().putVault(bearer, req) }.getOrElse {
            return if (it is HttpException && it.code() == 409) {
                val winner = fetchCurrentEnvAfter409(bearer) ?: return SyncOutcome.NetworkError(RuntimeException("no winner"))
                SyncOutcome.ConflictNeedManual(winner.version, expectedVersion)
            } else SyncOutcome.NetworkError(it)
        }
        return when (resp.code()) {
            200 -> {
                val body = resp.body()!!
                if (body.ok) SyncOutcome.Pushed(body.version!!)
                else SyncOutcome.NetworkError(RuntimeException(body.error ?: "put failed"))
            }
            409 -> {
                val winner = resp.body()?.current?.version
                    ?: fetchCurrentEnvAfter409(bearer)?.version
                    ?: return SyncOutcome.NetworkError(RuntimeException("409 missing current"))
                SyncOutcome.ConflictNeedManual(winner, expectedVersion)
            }
            401 -> SyncOutcome.AuthExpired("401")
            else -> SyncOutcome.NetworkError(
                RuntimeException("PUT /vault HTTP ${resp.code()} ${resp.errorBody()?.string()}")
            )
        }
    }

    private suspend fun fetchCurrentEnvAfter409(bearer: String) =
        runCatching { api.get().getVault(bearer).body() }.getOrNull()
}

File: flutter/lib/sync/sync_orchestrator.dart
import 'dart:convert';

import '../api/cloud_sync_api.dart';
import '../api/dto.dart';
import '../crypto/vault_crypto.dart';
import '../models/envelope.dart';
import '../models/sync_outcome.dart';
import '../models/vault.dart';
import '../session/vault_session.dart';
import '../storage/app_prefs.dart';
import '../storage/vault_blob_store.dart';

/// 云同步编排器 —— 与 packages/host/lib/cloud-sync.mjs 的 push / pull / confirmVersion
/// 语义对齐，也与 Android sync 模块的 SyncOrchestrator 一致。
///
/// 协议约定：云端存的是 envelope（PBKDF2+AES-GCM 密文），本应用本地是
/// 明文 vault。同步时用 [AppPrefs.cloudMasterPassword]（用户事先在 Settings
/// 里填好的"云密码"）seal / unseal。本类**不直接持有密码**：密码只在
/// `syncNow()` 调用前后从 prefs 读出、用完即丢。
///
/// 乐观并发流程：
///   1) GET /vault -> remoteVersion + remote envelope
///   2) remoteVersion > 本地 currentVersion -> 用云密码 unseal remote envelope
///      -> 成功则覆盖本地 + blobStore.adopt(vault)；失败则 ConflictNeedManual
///   3) 本地有改动 (dirty flag) 或本地版本 0 -> 用云密码 seal 本地 vault
///      -> PUT /vault (expectedVersion = lastReadVersion)
///        200 -> confirmVersion(v+1)
///        409 -> 读取 current envelope -> 用云密码 unseal -> 再 PUT 一次
///               expectedVersion=current.version
///                 仍 409 -> ConflictNeedManual 交给用户
class SyncOrchestrator {
  final CloudSyncApi api;
  final VaultBlobStore blobStore;
  final VaultSession session;
  final AppPrefs prefs;

  SyncOrchestrator({
    required this.api,
    required this.blobStore,
    required this.session,
    required this.prefs,
  });

  /// 触发一次云同步。返回 [SyncOutcome] 描述本次结果（参见
  /// `models/sync_outcome.dart`）。
  ///
  /// 调用前要求：
  ///   - 用户已经在 Settings 配好 serverUrl / cloud account
  ///     （即 [AppPrefs.sessionToken] 非空）；
  ///   - 用户已经在 Settings 填好 [AppPrefs.cloudMasterPassword]。
  /// 缺任一项会立即返回对应的 [SyncAuthExpired] / [SyncMissingCloudPassword]
  /// 让 UI 给出明确提示。
  Future<SyncOutcome> syncNow() async {
    final token = prefs.sessionToken;
    if (token == null || token.isEmpty) {
      return const SyncAuthExpired('not logged in');
    }
    final cloudPw = prefs.cloudMasterPassword;
    if (cloudPw == null) {
      return const SyncMissingCloudPassword();
    }
    final deviceId = prefs.deviceId;

    try {
      // 1) GET remote
      final getResult = await api.getVault();
      if (getResult.statusCode != 200 || !getResult.body.ok) {
        if (getResult.statusCode == 401) return const SyncAuthExpired('token rejected');
        return SyncNetworkError('GET /vault failed HTTP ${getResult.statusCode}');
      }
      final body = getResult.body;
      final remoteVersion = body.version;
      final remoteEnv = body.envelope;

      // Captured BEFORE any merge/overwrite below, because
      // `blobStore.overwriteWithRemote()` always resets the dirty flag —
      // we still need to know whether the pre-merge local vault had
      // changes worth re-uploading (otherwise a merged-in local edit could
      // silently never make it back to the relay).
      final wasDirtyBeforeMerge = blobStore.isDirty;
      var adoptedRemote = false;

      // 2) 有远程密文 & 更新 -> 拉取并与本地"合并"（按 entry id + updatedAt
      //    取更新的一份，tombstone 优先于旧副本），而不是整本覆盖。
      //
      //    IMPORTANT: this used to call `blobStore.overwriteWithRemote(
      //    decrypted, ...)` directly, which silently discarded any local
      //    edit that hadn't been pushed yet — the exact same class of bug
      //    fixed on the DSH host side in `cloudSyncNow()` (see
      //    packages/host/lib/index.mjs). Always merge, never overwrite.
      if (remoteEnv != null && remoteVersion > blobStore.currentVersion) {
        final decrypted = await _tryUnlock(remoteEnv, cloudPw);
        if (decrypted == null) {
          return const SyncDecryptFailed('remote vault sealed with different master password');
        }
        final merged = session.vault.mergeWith(decrypted);
        await blobStore.overwriteWithRemote(
          merged,
          remoteVersion,
          body.updatedAt ?? DateTime.now().millisecondsSinceEpoch,
        );
        await prefs.markSynced(remoteVersion, DateTime.now().millisecondsSinceEpoch);
        session.adopt(merged);
        adoptedRemote = true;
      }

      // 3) 本地有脏改动 / 从未提交过 -> push
      // 如果刚刚发生了合并，用"合并前"的 dirty 状态判断——overwriteWithRemote
      // 已经把 dirty 标志重置为 false，但合并进来的本地专属条目仍然需要
      // 重新上传，否则会在本地"看得到"但云端/其它设备永远收不到。
      final localDirty = adoptedRemote ? wasDirtyBeforeMerge : blobStore.isDirty;
      final nothingOnRemote = remoteVersion == 0 && remoteEnv == null;
      if (!localDirty && !nothingOnRemote) {
        return adoptedRemote
            ? SyncPulled(remoteVersion, true)
            : SyncPushed(remoteVersion);
      }

      final myEnv = await _sealLocal(cloudPw);
      if (myEnv == null) {
        return adoptedRemote ? SyncPulled(remoteVersion, true) : SyncPushed(remoteVersion);
      }

      final expectedVersion = remoteVersion > blobStore.currentVersion
          ? remoteVersion
          : blobStore.currentVersion;
      final firstPut = await _putVaultWithRetry(myEnv, expectedVersion, deviceId);
      if (firstPut is SyncPushed) {
        await blobStore.confirmVersion(firstPut.newVersion, DateTime.now().millisecondsSinceEpoch);
        await prefs.markSynced(firstPut.newVersion, DateTime.now().millisecondsSinceEpoch);
        return firstPut;
      }
      // 409 -> merge with the winner, then retry once (push the MERGED
      // snapshot, not the stale pre-conflict `myEnv` — pushing the old
      // local copy as-is would silently discard whatever the other device
      // just wrote, which is the same class of data-loss bug this whole
      // merge rework is fixing).
      if (firstPut is SyncConflictNeedManual) {
        final theirVersion = firstPut.theirVersion;
        final refetch = await _fetchCurrentEnvAfter409();
        final theirEnvDto = refetch?.envelope;
        if (theirEnvDto == null) {
          return SyncNetworkError('failed to refetch winner after 409');
        }
        final theirVault = await _tryUnlock(theirEnvDto, cloudPw);
        if (theirVault == null) {
          return const SyncDecryptFailed('remote vault sealed with different master password');
        }
        final merged = session.vault.mergeWith(theirVault);
        await blobStore.overwriteWithRemote(
          merged,
          theirVersion,
          DateTime.now().millisecondsSinceEpoch,
        );
        session.adopt(merged);
        final mergedEnv = await _sealLocal(cloudPw);
        if (mergedEnv == null) {
          return SyncPulled(theirVersion, true);
        }
        // retry PUT with the winner version
        final retry = await _putVaultWithRetry(mergedEnv, theirVersion, deviceId);
        if (retry is SyncPushed) {
          await blobStore.confirmVersion(retry.newVersion, DateTime.now().millisecondsSinceEpoch);
          await prefs.markSynced(retry.newVersion, DateTime.now().millisecondsSinceEpoch);
          return retry;
        }
        return retry;
      }
      return firstPut;
    } on CloudAuthExpiredException catch (e) {
      return SyncAuthExpired(e.message);
    } catch (t) {
      return SyncNetworkError(t);
    }
  }

  /// 用 [cloudPw] 试解 envelope；成功返回明文 Vault，失败返回 null。
  Future<Vault?> _tryUnlock(Envelope env, String cloudPw) async {
    try {
      final json = await VaultCrypto.unlock(envelope: env, masterPassword: cloudPw);
      return Vault.fromJson(jsonDecode(json) as Map<String, dynamic>);
    } on BadMasterPasswordException {
      return null;
    }
  }

  /// 把 session 当前内存 vault 封成 envelope。
  Future<Envelope?> _sealLocal(String cloudPw) async {
    final json = jsonEncode(session.vault.toJson());
    return VaultCrypto.seal(plaintextJson: json, masterPassword: cloudPw);
  }

  Future<SyncOutcome> _putVaultWithRetry(
    Envelope env,
    int expectedVersion,
    String deviceId,
  ) async {
    final req = PutVaultReq(envelope: env, expectedVersion: expectedVersion, deviceId: deviceId);
    try {
      final resp = await api.putVault(req);
      switch (resp.statusCode) {
        case 200:
          final body = resp.body;
          if (body.ok && body.version != null) return SyncPushed(body.version!);
          return SyncNetworkError(body.error ?? 'put failed');
        case 409:
          final winnerVersion = resp.body.current?.version ??
              (await _fetchCurrentEnvAfter409())?.version;
          if (winnerVersion == null) {
            return SyncNetworkError('409 missing current');
          }
          return SyncConflictNeedManual(winnerVersion, expectedVersion);
        case 401:
          return const SyncAuthExpired('401');
        default:
          return SyncNetworkError('PUT /vault HTTP ${resp.statusCode}');
      }
    } on CloudAuthExpiredException {
      final winner = await _fetchCurrentEnvAfter409();
      if (winner == null) return SyncNetworkError('no winner');
      return SyncConflictNeedManual(winner.version, expectedVersion);
    }
  }

  Future<GetVaultResp?> _fetchCurrentEnvAfter409() async {
    try {
      final result = await api.getVault();
      return result.body;
    } catch (_) {
      return null;
    }
  }
}